File: Leetcode/Lowest Common Ancestor of a Binary Tree/Lowest Common Ancestor of a Binary Tree.cpp
//Lca of nodes n1 and n2 is shared ancestor of nodes n1 and n2 which is located farthest from the roots

//In a tree we should always think for a particular node and recursion will take care of all nodes

//Here, the current node is root
// If root is Null then no LCA return NULL

// If root is either p or q then LCA is root
// Else if tell the left subtree to find LCA and right subtree to find LCA and store in left and right variables
//If left and right both are not null then root is LCA
//If left is NULL and right is NOT NULL then left subtree contains LCA so simply return left
//Vice Versa for right

//Else if left and right both are NULL return NULL

//Refer to code line by line to understand completely.

class Solution {
public:
    TreeNode* lca(TreeNode* root, TreeNode* p, TreeNode* q)
    {
        if(!root) return NULL;
        
        if(root->val==p->val or root->val==q->val)
            return root;
        
        TreeNode* left=lca(root->left,p,q);
        TreeNode* right=lca(root->right,p,q);
        
        if(left and right)
            return root;
        
        if(left)
            return left;
        
        if(right)
            return right;
        
        return NULL;
    }
    
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return lca(root,p,q);
    }
};

File: Leetcode/Lowest Common Ancestor of a Binary Tree/Lowest Common Ancestor of a Binary Tree.java
// PROBLEM STATEMENT:
// Given a binary tree, find the lowest common ancestor (LCA) of two given nodes in the tree.

// According to the definition of LCA on Wikipedia: “The lowest common ancestor is defined between two nodes p and q as the lowest node in T
//  that has both p and q as descendants (where we allow a node to be a descendant of itself).”

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // Applying the base condition
        if(root == null)
            return null;
        // If this condition is true then that root value will be the ancestor of p & q.
        if(root.val == p.val || root.val == q.val)
            return root;
        // Recursive Solution: apply Depth First Search. & traverse left subtree of the node and then right subtree of the node recursively.
        TreeNode l = lowestCommonAncestor(root.left,p,q);
        TreeNode r = lowestCommonAncestor(root.right,p,q);
        // (l != null && r != null ) found p and q in left and right subtree, return itself. 
        if(l != null && r != null)
            return root;
        // (l != null && r == null ) or (l == null || r != null) propogate it above.
        if(l != null || r == null)
            return l;
        if(l == null || r != null)
            return r;
        //  If you dont find neither p nor q in subtress, return null, since current node cannot be the LCA.
        return null;
        
    }
}